pkg/kubelet/cm/cpumanager/policy_static.go

```diff
@@ -571,10 +572,42 @@ func (p *staticPolicy) generateCPUTopologyHints(availableCPUs cpuset.CPUSet, reu
 	// to the minAffinitySize. Only those with an equal number of bits set (and
 	// with a minimal set of numa nodes) will be considered preferred.
 	for i := range hints {
+		if p.options.AlignBySocket && p.isHintSocketAligned(hints[i]) {
+			hints[i].Preferred = true
+			continue
+		}
 		if hints[i].NUMANodeAffinity.Count() == minAffinitySize {
 			hints[i].Preferred = true
 		}
 	}
 
 	return hints
 }
+
+// isHintSocketAligned function return true if numa nodes in hint are within one socket.
+func (p *staticPolicy) isHintSocketAligned(hint topologymanager.TopologyHint) bool {
+	numaNodesBitMask := hint.NUMANodeAffinity.GetBits()
+	return p.topology.CPUDetails.SocketsInNUMANodes(numaNodesBitMask...).Size() == 1
```

Comment:
Is it sufficient to just check `Size() == 1`?  What if `minAffinitySize` is greater than the number of numa nodes on a socket?

Should this really be something like:
```
minSockets := (minAffinitySize + numaNodesPerSocket - 1) / numaNodesPerSocket
return p.topology.CPUDetails.SocketsInNUMANodes(numaNodesBitMask...).Size() == minSockets
```

We should also add a test case for this.

Reply:
Can you give an example topology where above will apply?

Reply:
It is independent of the topology. It depends on the request size. Imagine a 2 socket machine with 4 NUMA nodes per socket and 20 CPUs per NUMA node. If a request comes in for 81 CPUs, this requires at least 5 NUMA nodes (meaning minAffinity will be 5) and (by definition) at least 2 sockets.

Reply:
The thought process i had is:
If allocation need to spread across two sockets, CPUs by definition are not socket aligned. Assumption is by socket alignment, we mean CPUs from one socket are preferred for allocation over cross-socket allocation.
In 2 socket architecture, With above logic, if allocation spreads two sockets then all hints become preferred and the whole point of finding optimized hint may become moot.

In my view, considering minSockets may be relevant for architecture where there are more than 2 sockets(viz 4 socket or 6 sockets etc) where we want the CPUs to be localized to let say 2 sockets.

Reply:
Intent is to have simplicity, in sense that CPU's will be considered aligned if they can come from one-socket. In my opinion, Making align-by-socket to consider CPUs as aligned within minAffinitySockets(2 or more ) may bring complexity whose performance returns may not be much. Since anyway allocation is anyway spreading across sockets

I could be wrong and open to suggestions.

Reply:
> If allocation need to spread across two sockets, CPUs by definition are not socket aligned.

This is true, but I don't see how that is relevant here. The point of the `align-by-socket` flag is not to try and force alignment to a single socket, but rather to generate hints at the socket boundary rather than the NUMA boundary.

If an allocation request has so many CPUs that it must *always* pull from more than one socket to satisfy its allocation (meaning, the topology dictates this, not the current allocation to other pods), then we need to generate a preferred hint that is representative of this.

Not doing so would make this flag inconsistent with its purpose -- i.e. to generate hints as if we were working at socket boundaries instead of NUMA boundaries.

Reply:
It's the reason we need to do this on the other side (instead of just assuming things always come from one socket):
```
	if p.options.AlignBySocket {
		socketBits := p.topology.CPUDetails.SocketsInNUMANodes(numaBits...).ToSliceNoSort()
		for _, socketID := range socketBits {
			alignedCPUs = alignedCPUs.Union(allocatableCPUs.Intersection(p.topology.CPUDetails.CPUsInSockets(socketID)))
		}
		return alignedCPUs
	}
```

Reply:
Fixed in latest commit